File: GitToolBox/src/main/java/zielu/gittoolbox/blame/calculator/IncrementalBlameBuilder.java
package zielu.gittoolbox.blame.calculator;

import com.intellij.execution.process.ProcessOutputType;
import com.intellij.openapi.util.Key;
import git4idea.commands.GitLineHandlerListener;
import git4idea.util.StringScanner;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * Parse output of incremental git blame.
 *
 * <p>Sample command line:
 * <pre>git blame --incremental -l -t -w --encoding=UTF-8 HEAD change-notes.html > blame-incremental.txt</pre>
 * </p>
 */
public class IncrementalBlameBuilder implements GitLineHandlerListener {
  private static final String FILENAME_TAG = "filename ";
  private static final String SUMMARY_TAG = "summary ";
  private static final String AUTHOR_TAG = "author ";
  private static final String AUTHOR_TIME_TAG = "author-time ";

  private final Map<String, CommitInfo> commits = new HashMap<>();
  private final List<Entry> entries = new ArrayList<>();
  private int lineCount = 0;

  private CommitInfo currentCommit;
  private Entry currentEntry;

  List<CommitInfo> buildLineInfos() {
    List<CommitInfo> lineInfos = new ArrayList<>(lineCount);
    for (int i = 0; i < lineCount; i++) {
      lineInfos.add(CommitInfo.NULL);
    }
    for (Entry entry : entries) {
      for (int i = 0; i < entry.numLines; i++) {
        int index = entry.resultLine - 1 + i;
        lineInfos.set(index, entry.commit);
      }
    }
    return lineInfos;
  }

  @Override
  public void onLineAvailable(String line, Key outputType) {
    if (ProcessOutputType.isStdout(outputType)) {
      parseLine(line);
    }
  }

  private void parseLine(String line) {
    if (currentEntry == null) {
      parseEntryStart(line);
    } else {
      if (line.startsWith(FILENAME_TAG)) {
        parseEntryEnd(line);
      } else {
        parseTagLine(line);
      }
    }
  }

  private void parseEntryStart(String line) {
    StringScanner scanner = new StringScanner(line);
    String commitHash = scanner.spaceToken();
    currentCommit = getCommit(commitHash);
    scanner.spaceToken(); //skip source line
    int resultLine = Integer.parseInt(scanner.spaceToken());
    int numLines = Integer.parseInt(scanner.line());
    currentEntry = new Entry(currentCommit, resultLine, numLines);
    lineCount = Math.max(lineCount, resultLine + (numLines - 1));
  }

  private CommitInfo getCommit(String commitHash) {
    return commits.computeIfAbsent(commitHash, CommitInfo::new);
  }

  private void parseEntryEnd(String line) {
    entries.add(currentEntry);
    currentEntry = null;
    currentCommit = null;
  }

  private void parseTagLine(String line) {
    if (line.startsWith(SUMMARY_TAG)) {
      currentCommit.setSummary(line.substring(SUMMARY_TAG.length()).trim());
    } else if (line.startsWith(AUTHOR_TAG)) {
      currentCommit.setAuthorName(line.substring(AUTHOR_TAG.length()).trim());
    } else if (line.startsWith(AUTHOR_TIME_TAG)) {
      currentCommit.setAuthorTime(Long.parseLong(line.substring(AUTHOR_TIME_TAG.length()).trim()));
    }
  }

  private static class Entry {
    private final CommitInfo commit;
    private final int resultLine;
    private final int numLines;

    private Entry(CommitInfo commit, int resultLine, int numLines) {
      this.commit = commit;
      this.resultLine = resultLine;
      this.numLines = numLines;
    }
  }
}

File: src/main/kotlin/zielu/gittoolbox/blame/calculator/IncrementalBlameBuilder.kt
package zielu.gittoolbox.blame.calculator

import com.intellij.execution.process.ProcessOutputType
import com.intellij.openapi.util.Key
import com.intellij.openapi.vcs.history.VcsRevisionNumber
import git4idea.commands.GitLineHandlerListener
import java.time.Instant
import java.time.ZoneId
import java.time.ZoneOffset
import java.util.HashMap
import java.util.regex.Matcher
import java.util.regex.Pattern
import kotlin.math.max

/**
 * Parse output of incremental git blame.
 *
 * <p>Sample command line:
 * <pre>git blame --incremental -l -t -w --encoding=UTF-8 HEAD change-notes.html > blame-incremental.txt</pre>
 * </p>
 *
 * @see (https://git-scm.com/docs/git-blame#_incremental_output)[https://git-scm.com/docs/git-blame#_incremental_output]
 */
internal class IncrementalBlameBuilder : GitLineHandlerListener {
  private val entries: MutableList<Entry> = ArrayList()
  private val commitBuilders: MutableMap<String, CommitInfo.Builder> = HashMap()
  private val commits: MutableMap<String, CommitInfo> = HashMap()
  private var currentEntry: Entry = NULL_ENTRY
  private var currentCommitBuilder = CommitInfo.Builder(VcsRevisionNumber.NULL)
  private var lineCount: Int = 0

  fun buildLineInfos(): List<CommitInfo> {
    val lineInfos: MutableList<CommitInfo> = ArrayList(lineCount)
    for (i in 0 until lineCount) {
      lineInfos.add(CommitInfo.NULL)
    }
    for (entry in entries) {
      for (i in 0 until entry.lineCount) {
        val index = entry.lineNumber - 1 + i
        lineInfos[index] = commits.computeIfAbsent(entry.commitHash) { hash: String ->
          commitBuilders[hash]!!.build()
        }
      }
    }
    return lineInfos
  }

  override fun onLineAvailable(line: String, outputType: Key<*>) {
    if (ProcessOutputType.isStdout(outputType)) {
      parseLine(line)
    }
  }

  private fun parseLine(line: String) {
    if (currentEntry === NULL_ENTRY) {
      parseEntryHeader(line)
    } else {
      if (line.startsWith(filenameTag)) {
        finishEntry()
      } else {
        parseTagLine(line)
      }
    }
  }

  private fun parseEntryHeader(line: String) {
    val matcher: Matcher = entryStartPattern.matcher(line)
    if (matcher.matches()) {
      parseEntryHeader(matcher)
    }
  }

  private fun parseEntryHeader(matcher: Matcher) {
    val commitHash = matcher.group(1)
    currentCommitBuilder = commitBuilders.computeIfAbsent(commitHash, CommitInfo::Builder)
    val lineNumber = matcher.group(2).toInt()
    val numLines = matcher.group(3).toInt()
    currentEntry = Entry(commitHash, lineNumber, numLines)
    lineCount = max(lineCount, lineNumber + (numLines - 1))
  }

  private fun finishEntry() {
    entries.add(currentEntry)
    currentEntry = NULL_ENTRY
    currentCommitBuilder = CommitInfo.Builder(VcsRevisionNumber.NULL)
  }

  private fun parseTagLine(line: String) {
    when {
      line.startsWith(summaryTag) -> currentCommitBuilder.summary(extractTagValue(summaryTag, line))
      line.startsWith(authorTag) -> currentCommitBuilder.authorName(extractTagValue(authorTag, line))
      line.startsWith(authorMailTag) -> currentCommitBuilder.authorEmail(extractEmailValue(authorMailTag, line))
      line.startsWith(authorTimeTag) -> currentCommitBuilder.authorTime(extractTimeValue(authorTimeTag, line))
      line.startsWith(authorTimeZoneTag) ->
        currentCommitBuilder.authorTimeOffset(
          extractTimeZoneValue(authorTimeZoneTag, line)
        )
    }
  }

  companion object {
    private val entryStartPattern = Pattern.compile("([a-zA-Z0-9]+)\\s+\\d+\\s+(\\d+)\\s+(\\d+)")
    private const val filenameTag = "filename "
    private const val summaryTag = "summary "
    private const val authorTag = "author "
    private const val authorTimeTag = "author-time "
    private const val authorTimeZoneTag = "author-tz "
    private const val authorMailTag = "author-mail "
    private val NULL_ENTRY = Entry("NULL", -1, 0)
  }
}

private data class Entry(
  val commitHash: String,
  val lineNumber: Int,
  val lineCount: Int
)

private fun extractTagValue(tag: String, line: String): String {
  return line.substring(tag.length).trim()
}

private fun extractEmailValue(tag: String, line: String): String {
  val value = extractTagValue(tag, line)
  return value.substring(1, value.length - 1)
}

private fun extractTimeValue(tag: String, line: String): Instant {
  val value = extractTagValue(tag, line)
  val secondsOfEpoch = value.toLong()
  return Instant.ofEpochSecond(secondsOfEpoch)
}

private fun extractTimeZoneValue(tag: String, line: String): ZoneId {
  val value = extractTagValue(tag, line)
  return ZoneOffset.of(value)
}
